docs/contributors/write-docstrings.rst

```diff
@@ -0,0 +1,178 @@
+
+.. code-block:: python
+
+    def function(arg1, arg2):
+        """Summary line.
+
+        Extended description of function.
+
+        Args:
+            arg1 (int): Description of arg1
+            arg2 (str): Description of arg2
+
+        Returns:
+            bool: Description of return value
+
+        Raises:
+            ValueError: If arg1 is equal to arg2
+
+        """
+        if arg1 == arg2:
+            raise ValueError('arg1 must not be equal to arg2')
+        return True
+
+If there are any examples of how to use the function, they should be listed in an ``Examples`` section. For example:
+
+.. code-block:: python
+
+    def function(arg1, arg2):
+        """Summary line.
+
+        Extended description of function.
+
+        Args:
+            arg1 (int): Description of arg1
+            arg2 (str): Description of arg2
+
+        Returns:
+            bool: Description of return value
+
+        Raises:
+            ValueError: If arg1 is equal to arg2
+
+        Notes:
+            Do not use this function if arg1 is equal to arg2.
+
+        Examples:
+            >>> function(1, 'a')
+            True
+
+        """
+        if arg1 == arg2:
+            raise ValueError('arg1 must not be equal to arg2')
+        return True
+
+The ``Examples`` section should contain code that can be executed by the user to demonstrate how to use the function. The code should be indented by four spaces.
+
+If there are any references that the user should be aware of, they should be listed in a ``References`` section. For example:
+
+.. code-block:: python
+
+    def function(arg1, arg2):
+        """Summary line.
+
+        Extended description of function.
+
+        Args:
+            arg1 (int): Description of arg1
+            arg2 (str): Description of arg2
+
+        Returns:
+            bool: Description of return value
+
+        Raises:
+            ValueError: If arg1 is equal to arg2
+
+        Notes:
+            Do not use this function if arg1 is equal to arg2.
+
+        Examples:
+            >>> function(1, 'a')
+            True
+
+        References:
+            - https://example.com
+            - https://example.com
+
+        """
+        if arg1 == arg2:
+            raise ValueError('arg1 must not be equal to arg2')
+
+
+So many rules, how can I check?
+-------------------------------
+
+There are many rules to follow when writing docstrings. How can you check that you are following them all?
+
+There's an easy way to check: use a tool called `pydocstyle <https://www.pydocstyle.org/en/stable/>`_.
+
+.. code-block:: bash
+
+    $ pip install pydocstyle
+
+``pydocstyle`` checks that your docstrings follow the `Google docstring style <https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html#example-google>`_.
+
+.. code-block:: bash
+
+    $ pydocstyle --convention=google --add-ignore=D212 gflownet/my_module.py
+    $ pydocstyle --convention=google --add-ignore=D212 gflownet/
```

Comment:
We could eventually add a check for that in the CI.
For now, it would probably not work given our mix of numpy/google docstrings but eventually...

Reply:
Yes absolutely. There's some work to do to adjust docstrings convention but that's a good goal to have.

Reply:
Just as a note, I didn't know about the various docstring conventions when I started building the repo. However, I have mostly followed, it seems, the Numpy style. So that vast majority of the repository resembles more the Numpy style than the Google style, although because I didn't stick to any rules, I am sure it also breaks much of the style convention.

Reply:
Yeah one day a style will have to be picked. In general Google is flatter, Numpy is more vertical. Also I think Numpy docstrings is more `rst`-sensitive than Google style but not sure. I have a visual preference for Google style when I read docstrings but it renders the same.